docs/design.md

```diff
@@ -0,0 +1,785 @@
+# Deckhand Design
+
+## Purpose
+
+Deckhand is a document-based configuration storage service built with
+auditability and validation in mind.
+
+## Essential Functionality
+
+* layering - helps reduce duplication in configuration while maintaining
+  auditability across many sites
+* substitution - provides separation between secret data and other
+  configuration data, while allowing a simple interface for clients
+* revision history - improves auditability and enables services to provide
+  functional validation of a well-defined collection of documents that are
+  meant to operate together
+* validation - allows services to implement and register different kinds of
+  validations and report errors
+
+## Documents
+
+All configuration data is stored entirely as structured documents, for which
+schemas must be registered.
+
+### Document Format
+
+The document format is modeled loosely after Kubernetes practices. The top
+level of each document is a dictionary with 3 keys: `schema`, `metadata`, and
+`data`.
+
+* `schema` - Defines the name of the JSON schema to be used for validation.
+  Must have the form: `<namespace>/<kind>/<version>`, where the meaning of
+  each component is:
+  * `namespace` - Identifies the owner of this type of document. The
+    values `deckhand` and `metadata` are reserved for internal use.
+  * `kind` - Identifies a type of configuration resource in the namespace.
+  * `version` - Describe the version of this resource, e.g. "v1".
+* `metadata` - Defines details that Deckhand will inspect and understand. There
+  are multiple schemas for this section as discussed below. All the various
+  types of metadata include a `name` field which must be unique for each
+  document `schema`.
+* `data` - Data to be validated by the schema described by the `schema`
+  field. Deckhand only interacts with content here as instructed to do so by
+  the `metadata` section. The form of this section is considered to be
+  completely owned by the `namespace` in the `schema`.
+
+#### Document Metadata
+
+There are 3 supported kinds of document metadata. Documents with `Document`
+metadata are the most common, and are used for normal configuration data.
+Documents with `Control` metadata are used to customize the behavior of
+Deckhand. Documents with `Tombstone` metadata are used to delete pre-existing
+documents with either `Document` or `Control` metadata.
+
+##### schema: metadata/Document/v1
+
+This type of metadata allows the following metadata hierarchy:
+
+* `name` - string, required - Unique within a revision for a given `schema`.
+* `storagePolicy` - string, required - Either `cleartext` or `encrypted`. If
+  `encyrpted` is specified, then the `data` section of the document will be
```

Comment:
I think either a centralized access policy approach or an individual
document-level approach could be fine.

Here's another example of access policy, with a slightly different format (and
slightly less flexible as it does not allow for layering or substitution):

```yaml
---
schema: deckhand/DocumentAccessPolicy/v1
metadata:
  schema: metadata/Control/v1
  name: document-access-policy
data:
  rules:
    - name: descriptive-name-01
      selector:
        foo: bar
      readRoles:
        - drydock
        - promenade
---
schema: deckhand/ValidationAccessPolicy/v1
metadata:
  schema: metadata/Control/v1
  name: validation-access-policy
data:
  armada-validation-01:
    writeRoles:
      - armada
---
schema: deckhand/ValidationPolicy/v1
metadata:
  schema: metadata/Control/v1
  name: validation-policy
data:
  readRoles:
    - shipyard
  validations:
    - name: same-format-as-before
...
```

On document read, you would iterate over the list of rules until finding a
successful match.  Any match would allow access.  Falling off the end of the
list with no hits would deny access.

`ValidationAccessPolicy` indicates which roles are allowed to post validation
updates for given revisions, and the new field in `ValidationPolicy` specifies
what roles are allowed to check the status of a validation policy and
transitively all its validations (this avoids keeping the two authorization
definitions consistent).

Here's an example of per-document policies for comparison:

```yaml
---
schema: deckhand/CertificateKey/v1
metadata:
  schema: metadata/Document/v1
  name: some-key
  storagePolicy: encrypted
  accessPolicy:
    readRoles:
      - armada
data: |
  SOME PRIVATE KEY
---
schema: deckhand/Validation/v1
metadata:
  schema: metadata/Control/v1
  name: armada-validation-01
data:
  writeRoles:
    - armada
---
schema: deckhand/ValidationPolicy/v1
metadata:
  schema: metadata/Control/v1
  name: validation-policy
data:
  readRoles:
    - shipyard
  validations:
    - name: same-format-as-before
...
```

In either case, I think we're only interested in specifying read access via
this metadata.  Write access is more complicated, because inserting a document
can have side effects due to the layering system; therefore, I think we should
configure write access to be limited to roles defined directly in deckhand
configuration.

I would think we would want to start with a centralized access policy and
consider adding per-document policy to override these if we find they are
needed.

Either of these approaches could be added in a backwards-compatible way.